test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,1374 @@
+    for _, case in pairs(cases) do
+      refresh(case)
+      local command = "%s/lines/LINES/g"
+
+      for i = 1, string.len(command) do
+        local c = string.sub(command, i, i)
+        execute("cnoremap ".. c .. " " .. c)
+      end
+
+      execute(command)
+      expect([[
+        Inc substitution on
+        two LINES
+        ]])
+      end
+  end)
+
+  it('work then mappings move the cursor', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap ,S LINES/<left><left><left><left><left><left>")
+
+      feed(":%s/lines/,Sor three <enter>")
+      expect([[
+        Inc substitution on
+        two or three LINES
+        ]])
+
+      execute("cnoremap ;S /X/<left><left><left>")
+      feed(":%s/;SI<enter>")
+      expect([[
+        Xnc substitution on
+        two or three LXNES
+        ]])
+
+      execute("cnoremap ,T //Y/<left><left><left>")
+      feed(":%s,TX<enter>")
+      expect([[
+        Ync substitution on
+        two or three LYNES
+        ]])
+
+      execute("cnoremap ;T s//Z/<left><left><left>")
+      feed(":%;TY<enter>")
+      expect([[
+        Znc substitution on
+        two or three LZNES
+        ]])
+      end
+  end)
+
+  it('work with a failing mapping', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x execute('bwipeout!')[-1].'x'")
+      source([[
+        func! DoIt()
+          bwipeout!
+        endfunc]])
+      --execute("cnoremap <expr> x DoIt()[-1].'x'")
+
+      feed(":%s/tw/tox<enter>")
+
+      -- error thrown b/c of the mapping
+      neq(nil, string.find(eval('v:errmsg'), '^E523:'))
+      -- the substitution after the error only works for ics=split/nosplit
+      -- which seems like the right thing to do in all cases, but we probably
+      -- don't want to change the default, so all in all this seems alright
+      if case == '' then
+        expect(default_text)
+      else
+        expect(string.gsub(default_text, "tw", "tox"))
+      end
+    end
+  end)
+
+  it('work when temporarily moving the cursor', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x cursor(1, 1)[-1].'x'")
+
+      feed(":%s/tw/tox/g<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+  it('work when a mapping disables incsub', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x execute('set incsubstitute=')[-1]")
+
+      feed(":%s/tw/toxa/g<enter>")
+      expect(string.gsub(default_text, "tw", "toa"))
+    end
+  end)
+
+  it('work with a complex mapping', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      source([[cnoremap x <C-\>eextend(g:, {'fo': getcmdline()})
+      \.fo<CR><C-c>:new<CR>:bw!<CR>:<C-r>=remove(g:, 'fo')<CR>x]])
+
+      feed(":%s/tw/tox")
+      feed("/<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
```

Comment:
Note that all string methods are string methods. You do not need to ever mention `string.*` in lua: `default_text:gsub('tw', 'tox')`.